<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Model-Validation:-The-Bias-Variance-Tradeoff-and-the-Train-Test-Split" data-toc-modified-id="Model-Validation:-The-Bias-Variance-Tradeoff-and-the-Train-Test-Split-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Model Validation: The Bias-Variance Tradeoff and the Train-Test Split</a></span></li><li><span><a href="#Objectives" data-toc-modified-id="Objectives-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Objectives</a></span></li><li><span><a href="#Motivation" data-toc-modified-id="Motivation-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Motivation</a></span></li><li><span><a href="#The-Bias-Variance-Tradeoff" data-toc-modified-id="The-Bias-Variance-Tradeoff-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>The Bias-Variance Tradeoff</a></span><ul class="toc-item"><li><span><a href="#A-Model-Example" data-toc-modified-id="A-Model-Example-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>A Model Example</a></span><ul class="toc-item"><li><span><a href="#Model-A" data-toc-modified-id="Model-A-4.1.1"><span class="toc-item-num">4.1.1&nbsp;&nbsp;</span>Model A</a></span></li><li><span><a href="#Model-B" data-toc-modified-id="Model-B-4.1.2"><span class="toc-item-num">4.1.2&nbsp;&nbsp;</span>Model B</a></span></li></ul></li><li><span><a href="#High-Bias-vs-High-Variance" data-toc-modified-id="High-Bias-vs-High-Variance-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>High-Bias vs High-Variance</a></span><ul class="toc-item"><li><span><a href="#Bias" data-toc-modified-id="Bias-4.2.1"><span class="toc-item-num">4.2.1&nbsp;&nbsp;</span>Bias</a></span></li><li><span><a href="#Variance" data-toc-modified-id="Variance-4.2.2"><span class="toc-item-num">4.2.2&nbsp;&nbsp;</span>Variance</a></span></li><li><span><a href="#Balancing-Bias-and-Variance" data-toc-modified-id="Balancing-Bias-and-Variance-4.2.3"><span class="toc-item-num">4.2.3&nbsp;&nbsp;</span>Balancing Bias and Variance</a></span></li></ul></li><li><span><a href="#Different-Models-for-King-County-Housing-Data" data-toc-modified-id="Different-Models-for-King-County-Housing-Data-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Different Models for King County Housing Data</a></span></li><li><span><a href="#🧠-Knowledge-Check" data-toc-modified-id="🧠-Knowledge-Check-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>🧠 Knowledge Check</a></span></li></ul></li><li><span><a href="#Train-Test-Split" data-toc-modified-id="Train-Test-Split-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Train-Test Split</a></span><ul class="toc-item"><li><span><a href="#Is-the-Model-Overfitting-or-Underfitting?" data-toc-modified-id="Is-the-Model-Overfitting-or-Underfitting?-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Is the Model Overfitting or Underfitting?</a></span></li><li><span><a href="#Exercise:-Name-that-Model!" data-toc-modified-id="Exercise:-Name-that-Model!-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Exercise: Name that Model!</a></span></li><li><span><a href="#Should-You-Ever-Fit-on-Your-Test-Set?" data-toc-modified-id="Should-You-Ever-Fit-on-Your-Test-Set?-5.3"><span class="toc-item-num">5.3&nbsp;&nbsp;</span>Should You Ever Fit on Your Test Set?</a></span></li><li><span><a href="#Train-Test-Split-Our-Earlier-Example" data-toc-modified-id="Train-Test-Split-Our-Earlier-Example-5.4"><span class="toc-item-num">5.4&nbsp;&nbsp;</span>Train-Test Split Our Earlier Example</a></span></li><li><span><a href="#Now-check-performance-on-test-data" data-toc-modified-id="Now-check-performance-on-test-data-5.5"><span class="toc-item-num">5.5&nbsp;&nbsp;</span>Now check performance on test data</a></span></li><li><span><a href="#🧠-Knowledge-Check" data-toc-modified-id="🧠-Knowledge-Check-5.6"><span class="toc-item-num">5.6&nbsp;&nbsp;</span>🧠 Knowledge Check</a></span></li><li><span><a href="#Same-Procedure-with-a-Polynomial-Model" data-toc-modified-id="Same-Procedure-with-a-Polynomial-Model-5.7"><span class="toc-item-num">5.7&nbsp;&nbsp;</span>Same Procedure with a Polynomial Model</a></span></li><li><span><a href="#Exercise" data-toc-modified-id="Exercise-5.8"><span class="toc-item-num">5.8&nbsp;&nbsp;</span>Exercise</a></span></li></ul></li><li><span><a href="#k-Fold-Cross-Validation:-Even-More-Rigorous-Validation" data-toc-modified-id="k-Fold-Cross-Validation:-Even-More-Rigorous-Validation-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>k-Fold Cross-Validation: Even More Rigorous Validation</a></span><ul class="toc-item"><li><span><a href="#Example" data-toc-modified-id="Example-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>Example</a></span><ul class="toc-item"><li><span><a href="#Simple-Model" data-toc-modified-id="Simple-Model-6.1.1"><span class="toc-item-num">6.1.1&nbsp;&nbsp;</span>Simple Model</a></span></li><li><span><a href="#More-Complex-Model" data-toc-modified-id="More-Complex-Model-6.1.2"><span class="toc-item-num">6.1.2&nbsp;&nbsp;</span>More Complex Model</a></span></li><li><span><a href="#Medium-Complexity-Model" data-toc-modified-id="Medium-Complexity-Model-6.1.3"><span class="toc-item-num">6.1.3&nbsp;&nbsp;</span>Medium-Complexity Model</a></span></li><li><span><a href="#Checking-Our-Models-Against-the-Holdout-Test-Set" data-toc-modified-id="Checking-Our-Models-Against-the-Holdout-Test-Set-6.1.4"><span class="toc-item-num">6.1.4&nbsp;&nbsp;</span>Checking Our Models Against the Holdout Test Set</a></span><ul class="toc-item"><li><span><a href="#Testing-Other-Models" data-toc-modified-id="Testing-Other-Models-6.1.4.1"><span class="toc-item-num">6.1.4.1&nbsp;&nbsp;</span>Testing Other Models</a></span></li></ul></li></ul></li></ul></li></ul></div>

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import PolynomialFeatures
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import cross_validate

# Model Validation: The Bias-Variance Tradeoff and the Train-Test Split

# Objectives

- **Explain** the bias-variance tradeoff and the correlative notions of underfit and overfit models
- **Describe** a train-test split and **explain** its purpose in the context of predictive statistics / machine learning
- **Explain** the algorithm of cross-validation

# Motivation

At this point, we have seen different ways to create models from our data through different linear regression techniques. That's great but just like a student practicing for a big end-of-the-year test, we want to make sure our _models_ are ready to predict on data it hasn't seen yet. We want to know if the model we made is ready to make predictions for data in the "wild". 

Usually, when our model is ready to be used in the "real world" we refer to this as putting our model into **production** or **deploying** our model. The data it will use to make predictions will be data its never seen before. Similar to parents sending a once teenager into the world to make it on their own, we want to make sure our model is ready for the world of new data!

But you might be thinking, how do I make sure my model I've been cultivating with my own data is ready? This is where we ***model validation*** techniques to ensure our model can generalize to data it hasn't directly seen before. Going back to our analogy of a teenager ready to leave the home, we want our teenager (model) be informed so it's not naive but also flexible enough to adjust to new situations.

We'll go over how to ensure our model is ready, but first we have to discuss how our model can make errors in the context of the **bias-variance tradeoff**.

# The Bias-Variance Tradeoff

## A Model Example

Typically we'll talk about a model as how _complex_ it is in making predictions.

Let's take a look at this data with just one feature and a target:

<!--TODO: Replace with a dataset and code -->
![](https://camo.githubusercontent.com/36a1cb13983f39fc58ecdfffb415b2258e73ba1c/68747470733a2f2f6769746875622e636f6d2f6c6561726e2d636f2d73747564656e74732f6473632d322d32342d30372d626961732d76617269616e63652d74726164652d6f66662d6f6e6c696e652d64732d73702d3030302f7261772f6d61737465722f696e6465785f66696c65732f696e6465785f375f312e706e67)

We can probably picture how a good model will fit to this data. Let's look at a couple models and discuss how they're making mistakes.

### Model A

<!--TODO: Replace with code to implement simple model -->

![](img/model_simple.png)

What do we observe here? How would you describe where the model is failing?

### Model B

<!--TODO: Replace with code to implement complex (overfitting) model -->

![](img/model_complex.png)

What do we observe here? How would you describe where the model is failing?

## High-Bias vs High-Variance

We can break up how the model makes mistakes (the error) by saying there are three parts:

- Error inherent of the data (noise): **irreducible error**
- Error from being not capturing patterns (too simple): **bias**
- Error from using patterns in the data but don't generalize well (too complex): **variance**

We can summarize this in an equation for the _mean squared error_ (MSE):

$MSE = Bias(\hat{y})^2 + Var(\hat{y}) + \sigma^2$

![optimal](img/optimal_bias_variance.png)
http://scott.fortmann-roe.com/docs/BiasVariance.html

### Bias

**High-bias** algorithms tend to be less complex, with simple or rigid underlying structure.

![](img/noisy-sine-linear.png)

+ They train models that are consistent, but inaccurate on average.
+ These include linear or parametric algorithms such as regression and naive Bayes.
+ The following sorts of difficulties could lead to high bias:
  - We did not include the correct predictors
  - We did not take interactions into account
  - We missed a non-linear (polynomial) relationship

      
High-bias models are generally **underfit**: The models have not picked up enough of the signal in the data. And so even though they may be consistent, they don't perform particularly well on the initial data, and so they will be consistently inaccurate.

### Variance

On the other hand, **high-variance** algorithms tend to be more complex, with flexible underlying structure.

![](img/noisy-sine-decision-tree.png)

+ They train models that are accurate on average, but inconsistent.
+ These include non-linear or non-parametric algorithms such as decision trees and nearest-neighbor models.
+ The following sorts of difficulties could lead to high variance:
  - We included an unreasonably large number of predictors;
  - We created new features by squaring and cubing each feature.

High variance models are **overfit**: The models have picked up on the noise as well as the signal in the data. And so even though they may perform well on the initial data, they will be inconsistently accurate on new data.

### Balancing Bias and Variance

While we build our models, we have to keep this relationship in mind.  If we build complex models, we risk overfitting our models.  Their predictions will vary greatly when introduced to new data.  If our models are too simple, the predictions as a whole will be inaccurate.   

![](img/noisy-sine-third-order-polynomial.png)

The goal is to build a model with enough complexity to be accurate, but not too much complexity to be erratic.

![](img/target.png)

## 🧠 Knowledge Check

![which_model](img/which_model_is_better_2.png)

# Train-Test Split

It is hard to know if your model is too simple or complex by just using it on training data.

We can _hold out_ part of our training sample, use it as a test sample, and then use it to monitor our prediction error.

This allows us to evaluate whether our model has the right balance of bias/variance. 

<img src='img/testtrainsplit.png' width =550 />

* **training set** —a subset to train a model.
* **test set**—a subset to test the trained model.

In [38]:
df = pd.read_csv('data/king_county.csv', index_col='id')

y = df.price
X = df[['bedrooms', 'sqft_living']]

#If you don't specify test_size it will be 25% 
X_train, X_test, y_train, y_test = train_test_split(X, 
                                                    y,
                                                    test_size=None,
                                                    random_state=42 #to avoid data leakage - so observation 
                                                                #from train may leak into test if you 
                                                                #since train test split is randomized
                                                                #want train and test reproducable
)



In [39]:
display(X_train.head())
display(X_test.head())

bedrooms  sqft_living
id                               
9117000170         4         1810
6700390210         3         1600
7212660540         4         1720
8562780200         2         1240
7760400350         3         1280

bedrooms  sqft_living
id                               
2591820310         4         2070
7974200820         5         2900
7701450110         4         3770
9522300010         3         4560
9510861140         3         2550

In [40]:
print(X_train.shape)
print(X_test.shape)

print(X_train.shape[0] == y_train.shape[0])
print(X_test.shape[0] == y_test.shape[0])

(16209, 2)
(5404, 2)
True
True


## Is the Model Overfitting or Underfitting?

If our model is not performing well on the training  data, we are probably underfitting it.  

To know if our  model is overfitting the data, we need  to test our model on unseen data. 
We then measure our performance on the unseen data. 

If the model performs significantly worse on the  unseen data, it is probably  overfitting the data.

<img src='https://developers.google.com/machine-learning/crash-course/images/WorkflowWithTestSet.svg' width=500/>

## Exercise: Name that Model!

Consider the following scenarios and describe them according to bias and variance. There are four possibilities and add whether it is potentially overfitting/underfitting or if it is a "sweet spot":

- a. The model has low bias and high variance.
- b. The model has high bias and low variance.
- c. The model has both low bias and low variance.
- d. The model has both high bias and high variance.

**Scenario 1**: The model has a low RMSE on training and a low RMSE on test.
<details>
    <summary> Answer
    </summary>
    c. The model has both low bias and low variance.
    </details>

**Scenario 2**: The model has a high $R^2$ on the training set, but a low $R^2$ on the test.
<details>
    <summary> Answer
    </summary>
    a. The model has low bias and high variance.
    </details>

**Scenario 3**: The model performs well on data it is fit on and well on data it has not seen.
<details>
    <summary> Answer
    </summary>
    c. The model has both low bias and low variance.
    </details>
  

**Scenario 4**: The model has a low $R^2$ on training but high on the test set.
<details>
    <summary> Answer
    </summary>
    d. The model has both high bias and high variance.
    </details>

**Scenario 5**: The model leaves out many of the meaningful predictors, but is consistent across samples.
<details>
    <summary> Answer
    </summary>
    b. The model has high bias and low variance.
    </details>

**Scenario 6**: The model is highly sensitive to random noise in the training set.
<details>
    <summary> Answer
    </summary>
    a. The model has low bias and high variance.
    </details>

## Should You Ever Fit on Your Test Set?  

![no](https://media.giphy.com/media/d10dMmzqCYqQ0/giphy.gif)

**Never fit on test data.** If you are seeing surprisingly good results on your evaluation metrics, it might be a sign that you are accidentally training on the test set.

## Train-Test Split Our Earlier Example

Let's go back to our KC housing data without the polynomial transformation.

In [41]:
df = pd.read_csv('data/king_county.csv', index_col='id')
df.head()

price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
id                                                                         
7129300520  221900.0         3       1.00         1180      5650     1.0   
6414100192  538000.0         3       2.25         2570      7242     2.0   
5631500400  180000.0         2       1.00          770     10000     1.0   
2487200875  604000.0         4       3.00         1960      5000     1.0   
1954400510  510000.0         3       2.00         1680      8080     1.0   

            waterfront  view  condition  grade  sqft_above  sqft_basement  
id                                                                         
7129300520           0     0          3      7        1180              0  
6414100192           0     0          3      7        2170            400  
5631500400           0     0          3      6         770              0  
2487200875           0     0          5      7        1050            910  
1954400510           0     0          3      8        1680              0

Now, we create a train-test split via the `sklearn.model_selection` package.

In [42]:
y = df.price
X = df[['bedrooms', 'sqft_living']]

# Here is the convention for a traditional train-test split
X_train, X_test, y_train, y_test = train_test_split(X, y,
                                                    test_size=0.2,
                                                    random_state=42)

In [43]:
# Instanstiate your linear regression object
lr = LinearRegression()

In [44]:
# fit the model on the training set
lr.fit(X_train, y_train)

LinearRegression()

In [45]:
# Check the R^2 of the training data
lr.score(X_train, y_train)

0.5059385369004967

In [46]:
lr.coef_

array([-55228.54258621,    312.15699349])

A .506 R-squared reflects a model that explains about half of the total variance in the data. 

## Now check performance on test data

Next, we test how well the model performs on the unseen test data. Remember, we do not fit the model again. The model has calculated the optimal parameters learning from the training set.  

In [47]:
lr.score(X_test, y_test)

0.5093504668464703

## 🧠 Knowledge Check

How would you describe the bias of the model based on the above training $R^2$?

<details>
    <summary> Answer
    </summary>
    The difference between the train and test scores is low.
    </details>

What does that indicate about variance?

## Same Procedure with a Polynomial Model

In [48]:
df = pd.read_csv('data/king_county.csv', index_col='id')
df.head()

price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
id                                                                         
7129300520  221900.0         3       1.00         1180      5650     1.0   
6414100192  538000.0         3       2.25         2570      7242     2.0   
5631500400  180000.0         2       1.00          770     10000     1.0   
2487200875  604000.0         4       3.00         1960      5000     1.0   
1954400510  510000.0         3       2.00         1680      8080     1.0   

            waterfront  view  condition  grade  sqft_above  sqft_basement  
id                                                                         
7129300520           0     0          3      7        1180              0  
6414100192           0     0          3      7        2170            400  
5631500400           0     0          3      6         770              0  
2487200875           0     0          5      7        1050            910  
1954400510           0     0          3      8        1680              0

In [49]:
poly_2 = PolynomialFeatures(4)

X_poly = pd.DataFrame(
            poly_2.fit_transform(df.drop('price', axis=1))
                      )

y = df.price
X_poly.head()

0     1     2       3        4     5     6     7     8     9     ...  \
0   1.0   3.0  1.00  1180.0   5650.0   1.0   0.0   0.0   3.0   7.0  ...   
1   1.0   3.0  2.25  2570.0   7242.0   2.0   0.0   0.0   3.0   7.0  ...   
2   1.0   2.0  1.00   770.0  10000.0   1.0   0.0   0.0   3.0   6.0  ...   
3   1.0   4.0  3.00  1960.0   5000.0   1.0   0.0   0.0   5.0   7.0  ...   
4   1.0   3.0  2.00  1680.0   8080.0   1.0   0.0   0.0   3.0   8.0  ...   

         1355          1356          1357          1358          1359  \
0         0.0  1.150122e+10  0.000000e+00  0.000000e+00  0.000000e+00   
1   7840000.0  7.152819e+10  1.318492e+10  2.430400e+09  4.480000e+08   
2         0.0  2.739198e+09  0.000000e+00  0.000000e+00  0.000000e+00   
3  40576900.0  8.103375e+09  7.022925e+09  6.086535e+09  5.274997e+09   
4         0.0  3.793306e+10  0.000000e+00  0.000000e+00  0.000000e+00   

           1360          1361          1362          1363          1364  
0  1.938778e+12  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  
1  2.217374e+13  4.087325e+12  7.534240e+11  1.388800e+11  2.560000e+10  
2  3.515304e+11  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  
3  1.215506e+12  1.053439e+12  9.129802e+11  7.912496e+11  6.857496e+11  
4  7.965942e+12  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  

[5 rows x 1365 columns]

In [50]:
X_train, X_test, y_train, y_test = train_test_split(X_poly, y,
                                                    test_size=0.2,
                                                    random_state=42)
lr_poly = LinearRegression()

# Always fit on the training set
lr_poly.fit(X_train, y_train)

lr_poly.score(X_train, y_train)

0.7025512345850116

In [51]:
lr_poly.score(X_test, y_test)

-54.536720695314614

## Exercise

**[This post about scaling and data leakage](https://datascience.stackexchange.com/questions/38395/standardscaler-before-and-after-splitting-data) explains that if you are going to scale your data, you should only train your scaler on the training data to prevent data leakage.**  

Perform the same train-test split as shown above for the simple model, but now scale your data appropriately.  

In [52]:
np.random.seed(42)

y = df.price
X = df[['bedrooms', 'sqft_living']]

# Train test split with random_state=42 and test_size=0.2

X_train, X_test, y_train, y_test = train_test_split(X, y,
                                                    test_size=0.2,
                                                    random_state=42)

# Scale appropriately on TRAINING DATA 

ss = StandardScaler()
ss.fit(X_train)
X_preds_st_scaled = ss.transform(X_train)

lr = LinearRegression()

# fit and score the model (checkout the test set if there is time)
lr.fit(X_preds_st_scaled, y_train)
lr.coef_
lr.score(X_preds_st_scaled, y_train)  #this scores should be the same as the first train-split example above 


0.5059385369004967

# k-Fold Cross-Validation: Even More Rigorous Validation  

Our goal of using a test set is to simulate what happens when our model attempts predictions on data it's never seen before. But there's always a chance our model will by chance perform well on the test set.

This is where we could use a more rigorous validation method and so we turn to **k-fold cross-validation**.

![kfolds](img/k_folds.png)

[image via sklearn](https://scikit-learn.org/stable/modules/cross_validation.html)

In this process, we split the dataset into a train set and holdout test sets like usual by performing a shuffling train-test split on the train set.  

We then do $k$-number of _folds_ of the training data. This means we divide the training set into different sections or folds. We then take turns on using each fold as a **validation set** (or **dev set**) and train on the larger fraction. Then we calculate a validation score from the validation set the model has never seen. We repeat this process until each fold has served as a validation set.

This process allows us to try out training our model and check to see if it is likely to overfit or underfit without touching the holdout test data set.

If we think the model is looking good according to our cross-validation using the training data, we retrain the model using all of the training data. Then we can do one final evaluation using the test data. 

It's important that we hold onto our test data until the end and refrain from making adjustments to the model based on the test results.

## Example

In [53]:
X = df.drop('price', axis=1)
y = df.price

In [54]:
# Let's create our holdout test
X_train, X_test, y_train, y_test = train_test_split(
                                                X,
                                                y,
                                                test_size=0.2,
                                                random_state=42
)

### Simple Model

In [19]:
model_simple = LinearRegression()
scores_simple = cross_validate(
                    model_simple, X_train, y_train, cv=5, 
                    return_train_score=True
)
scores_simple

{'fit_time': array([0.04891825, 0.05092692, 0.03490686, 0.06176019, 0.0385468 ]),
 'score_time': array([0.00161886, 0.00166106, 0.00126314, 0.00276089, 0.00316405]),
 'test_score': array([0.58939675, 0.59024623, 0.62247064, 0.61815269, 0.58479733]),
 'train_score': array([0.60608408, 0.60542007, 0.59746541, 0.59886644, 0.60586444])}

In [20]:
# Mean train r_2
np.mean(scores_simple['train_score']), np.std(scores_simple['train_score'])

(0.6027400889129975, 0.0037670560670530983)

In [21]:
# Mean test r_2
np.mean(scores_simple['test_score']), np.std(scores_simple['test_score'])

(0.6010127304502677, 0.01592487173906749)

In [22]:
# Fit on all the training data
model_simple.fit(X_train, y_train)
model_simple.score(X_train, y_train)

0.602533379449363

### More Complex Model

In [23]:
# Test out our polynomial model
poly_3 = PolynomialFeatures(3)
X_poly3 = poly_3.fit_transform(X_train)

In [24]:
model_poly3 = LinearRegression()
scores_complex3 = cross_validate(
                        model_poly3, X_poly3, y_train, cv=5, 
                        return_train_score=True
)
scores_complex3

{'fit_time': array([2.31131625, 2.31693411, 2.2756958 , 2.03604007, 1.91766119]),
 'score_time': array([0.00127077, 0.00098705, 0.00105214, 0.00099111, 0.00115585]),
 'test_score': array([ 0.68876959,  0.61665975,  0.38541365,  0.62112617, -0.25850185]),
 'train_score': array([0.6928725 , 0.70070931, 0.71077313, 0.71305421, 0.71916595])}

In [25]:
# Mean train r_2
np.mean(scores_complex3['train_score']), np.std(scores_complex3['train_score']) 

(0.7073150187567784, 0.009354819939372282)

In [26]:
# Mean test r_2
np.mean(scores_complex3['test_score']), np.std(scores_complex3['test_score'])

(0.41069346181153943, 0.3499976571854969)

In [27]:
# Fit on all the training data
model_poly3.fit(X_poly3, y_train)
model_poly3.score(X_poly3, y_train)

0.7033589903208285

### Medium-Complexity Model

In [28]:
# Test out our polynomial model
poly_2 = PolynomialFeatures(2)
X_poly2 = poly_2.fit_transform(X_train)

In [29]:
model_poly2 = LinearRegression()
scores_complex2 = cross_validate(
                        model_poly2, X_poly2, y_train, cv=5, 
                        return_train_score=True
)
scores_complex2

{'fit_time': array([0.37866497, 0.67976403, 0.63277388, 0.38723993, 0.38703489]),
 'score_time': array([0.00062299, 0.00120401, 0.00067616, 0.0006609 , 0.00058317]),
 'test_score': array([0.71440082, 0.5886194 , 0.66306261, 0.69362487, 0.63191255]),
 'train_score': array([0.68806328, 0.69353646, 0.69956413, 0.69467419, 0.70768282])}

In [30]:
# Mean train r_2
np.mean(scores_complex2['train_score']), np.std(scores_complex2['train_score']) 

(0.696704175441582, 0.006595914048353439)

In [31]:
# Mean test r_2
np.mean(scores_complex2['test_score']), np.std(scores_complex2['test_score'])

(0.6583240514191866, 0.04465313126535342)

In [32]:
model_poly2.fit(X_poly2, y_train)
model_poly2.score(X_poly2, y_train)

0.6957552617273535

### Checking Our Models Against the Holdout Test Set

Once we have an acceptable model, we train our model on the entire training set and score on the test to validate.

In [33]:
best_model = model_poly2

In [34]:
# Remember we have to transform X_test in the same way
best_model.score(
    poly_2.transform(X_test),
    y_test
)

0.678605344170726

#### Testing Other Models

In [35]:
# Simple model
model_simple.score(
    X_test,
    y_test
)

0.6150749892583851

In [36]:
# Remember we have to transform X_test in the same way for this model
model_poly3.score(
    poly_3.transform(X_test),
    y_test
)

0.6042672193696808